data_steward/cdr_cleaner/cleaning_rules/update_invalid_zip_codes.py

```diff
@@ -0,0 +1,197 @@
+"""
+Sandbox and update invalid zip codes found in the observation table.
+
+Original Issues: DC-1633, DC-1645
+
+The intent of this cleaning rule is to remove any leading/trailing whitespace in the zip code string then sandbox and
+update any invalid zip code in the observation table. A zip code is considered invalid if it:
+        Is less than 5 digits in length
+        Is alpha-numeric
+        Does not match any zip3 code in the master zip3 lookup table
+If zip code is deemed invalid the record is sandboxed and updated to have the following information:
+        value_as_string and value_source_value = 'Response removed due to invalid value'
+        value_as_number = 0
+        value_source_concept_id = 2000000010
+"""
+
+# Python imports
+import logging
+
+# Project imports
+from utils import pipeline_logging
+from constants.cdr_cleaner import clean_cdr as cdr_consts
+from common import JINJA_ENV, OBSERVATION, PIPELINE_TABLES, ZIP3_LOOKUP
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+
+LOGGER = logging.getLogger(__name__)
+
+ZIPS_WITH_WHITESPACE_SANDBOX = 'dc1633_zips_with_whitespace'
+
+# Creates sandbox that contains any zips that have leading/trailing whitespace
+SANDBOX_ZIPS_WITH_WHITESPACE = JINJA_ENV.from_string("""
+CREATE OR REPLACE TABLE `{{project_id}}.{{sandbox_dataset_id}}.{{whitespace_sandbox}}` AS
+SELECT * FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}`
+WHERE observation_source_concept_id = 1585250 
+AND REGEXP_CONTAINS(value_as_string, ' ')
+""")
+
+# Creates sandbox that contains all invalid zip codes which are deemed invalid because they:
+# 1. Are not 5 digits in length
+# 2. Are are alpha-numeric
+# 3. Do not match a zip3 code in the master zip3 lookup table
+SANDBOX_INVALID_ZIP_CODES = JINJA_ENV.from_string("""
+CREATE OR REPLACE TABLE `{{project_id}}.{{sandbox_dataset_id}}.{{sandbox_table}}` AS (
+-- Selects all zips that are less than 5 digits in length and/or alpha-numeric --
+SELECT * FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}`
+WHERE observation_source_concept_id = 1585250 
+AND (LENGTH(value_as_string) < 5 
+OR NOT REGEXP_CONTAINS(value_as_string, r'^[0-9]{5}(?:-[0-9]{4})?$'))
+-- Using union distinct to prevent duplicates --
+UNION DISTINCT (
+-- Selects all zips that do not match one in the master zip3 lookup table --
+SELECT o.* FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` o
+LEFT JOIN `{{project_id}}.{{pipeline_tables}}.{{zip3_lookup}}` z
+ON SUBSTR(o.value_as_string, 1, 3) = CAST(z.zip3 AS STRING)
+WHERE (observation_source_concept_id = 1585250 AND z.zip3 IS NULL)))
+""")
+
+CLEAN_ZIPS_OF_WHITESPACE = JINJA_ENV.from_string("""
+UPDATE `{{project_id}}.{{dataset_id}}.{{obs_table}}` SET
+value_as_string = TRIM(value_as_string)
+WHERE person_id IN (
```

Comment:
Right now, this is modifying all observations that the person has, rather than just the zip code observation. Changing the `WHERE` clause to the following should fix it:

```
WHERE observation_id IN (
SELECT observation_id FROM `{{project_id}}.{{sandbox_dataset_id}}.{{whitespace_sandbox}}`)
```

Reply:
Nice catch!